python/hail/linalg/blockmatrix.py

```diff
@@ -0,0 +1,483 @@
+        """
+
+        return BlockMatrix(self._jbm.filter(jarray(Env.jvm().long, rows_to_keep),
+                                            jarray(Env.jvm().long, cols_to_keep)))
+
+
+    def to_numpy(self):
+        """Create a `NumPy ndarray
+        <https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html>`__
+        by collecting the block matrix.
+
+        Warning
+        -------
+
+        This method is not efficient for large matrices.
+
+        Returns
+        -------
+        NumPy ``ndarray``
+            ``ndarray`` with shape ``(n_rows, n_cols)``.
+        """
+
+        return numpy_from_breeze(self._jbm.toBreezeMatrix())
+
+    @property
+    def T(self):
+        """Matrix transpose.
+
+        Returns
+        -------
+        :class:`.LocalMatrix`
+        """
+
+        return BlockMatrix(self._jbm.transpose())
+
+    def cache(self):
+        """Cache the RDD of blocks.
+
+        Returns
+        -------
+        :class:`.BlockMatrix`
+        """
+
+        return BlockMatrix(self._jbm.cache())
+
+    @typecheck_method(storage_level=storage_level)
+    def persist(self, storage_level='MEMORY_AND_DISK'):
+        """Persist the RDD of blocks.
+
+        Parameters
+        ----------
+        storage_level: :str:
+          Storage level for persistence.
+
+        Returns
+        -------
+        :class:`.BlockMatrix`
+        """
+
+        return BlockMatrix(self._jbm.persist(storage_level))
+
+    def unpersist(self):
+        """Unpersist the RDD of blocks.
+
+        Returns
+        -------
+        :class:`.BlockMatrix`
+        """
+
+        return BlockMatrix(self._jbm.unpersist())
+
+    @typecheck_method(b=block_matrix_type)
+    def __add__(self, b):
+        """Addition: A + B.
+
+        Parameters
+        ----------
+
+        b: :class:`BlockMatrix`
+
+        Returns
+        -------
+        :class:`.BlockMatrix`
+        """
+
+        return BlockMatrix(self._jbm.add(b._jbm))
+
+    @typecheck_method(b=block_matrix_type)
+    def __sub__(self, b):
+        """Subtraction: A - B.
+
+        Parameters
+        ----------
+
+        b: :class:`BlockMatrix`
+
+        Returns
+        -------
+        :class:`.BlockMatrix`
+        """
+
+        return BlockMatrix(self._jbm.subtract(b._jbm))
+
+    @typecheck_method(b=numeric)
+    def __mul__(self, b):
+        """Scalar multiplication: A * b
+
+        Parameters
+        ----------
+
+        b: :obj:`int` or :obj:`float`
+
+        Returns
+        -------
+        :class:`.BlockMatrix`
+        """
+
+        return BlockMatrix(self._jbm.scalarMultiply(float(b)))
+
+    @typecheck_method(b=numeric)
+    def __truediv__(self, b):
+        """Scalar division: A / b
+
+        Parameters
+        ----------
+
+        b: :obj:`int` or :obj:`float`
+
+        Returns
+        -------
+        :class:`.BlockMatrix`
+        """
+
+        return self * (1. / b)
+
+    @typecheck_method(b=block_matrix_type)
+    def dot(self, b):
```

Comment:
I agree, this is better off as matmul, as it will always return a matrix, unlike numpy dot. We will want to keep dot around on ndarray for the more general operations.